pkg/virt-api/validating-webhook/validating-webhook.go

```diff
@@ -0,0 +1,406 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package validating_webhook
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	v1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	cloudInitMaxLen = 2048
+)
+
+func getAdmissionReview(r *http.Request) (*v1beta1.AdmissionReview, error) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		return nil, fmt.Errorf("contentType=%s, expect application/json", contentType)
+	}
+
+	ar := &v1beta1.AdmissionReview{}
+	err := json.Unmarshal(body, ar)
+	return ar, err
+}
+
+func toAdmissionResponse(err error) *v1beta1.AdmissionResponse {
+	log.Log.Reason(err).Error("admitting vms")
+	return &v1beta1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+			Code:    http.StatusUnprocessableEntity,
+		},
+	}
+}
+
+type admitFunc func(*v1beta1.AdmissionReview) *v1beta1.AdmissionResponse
+
+func serve(resp http.ResponseWriter, req *http.Request, admit admitFunc) {
+	response := v1beta1.AdmissionReview{}
+	review, err := getAdmissionReview(req)
+
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	reviewResponse := admit(review)
+	if reviewResponse != nil {
+		response.Response = reviewResponse
+		response.Response.UID = review.Request.UID
+	}
+	// reset the Object and OldObject, they are not needed in a response.
+	review.Request.Object = runtime.RawExtension{}
+	review.Request.OldObject = runtime.RawExtension{}
+
+	responseBytes, err := json.Marshal(response)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed json encode webhook response")
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if _, err := resp.Write(responseBytes); err != nil {
+		log.Log.Reason(err).Errorf("failed to write webhook response")
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	resp.WriteHeader(http.StatusOK)
+}
+
+func validateDisks(disks []v1.Disk) []error {
+	errors := []error{}
+	for idx, disk := range disks {
```

Comment:
Could you check that the disk names are unique?

Reply:
Can we enfore a limit?

Reply:
yup, done. I set the array limit to 256